### PR TITLE
Add Uncork API Gateway support

### PR DESCRIPTION
**Pull request by edorofeev**
Improvements to make this middleware working as for current crawlera API and as for API gateway.

### CODE REVIEW

File: crawlera_fetch/middleware.py [hunk 1]
```diff
@@ -14,6 +14,8 @@
 from scrapy.spiders import Spider
 from scrapy.utils.reqser import request_from_dict, request_to_dict
 from w3lib.http import basic_auth_header
+import base64
+import binascii
```
Comment (elacuesta) [inline crawlera_fetch/middleware.py]: Could you move these stdlib imports up, as per [PEP8?](https://www.python.org/dev/peps/pep-0008/#imports)

File: crawlera_fetch/middleware.py [hunk 2]
```diff
@@ -178,19 +181,23 @@ def process_response(self, request: Request, response: Response, spider: Spider)
...
+        request_id = json_response.get("id") or json_response.get("uncork_id")
+        if server_error:
+            message = json_response.get("body") or json_response.get("message")
             self.stats.inc_value("crawlera_fetch/response_error")
-            self.stats.inc_value("crawlera_fetch/response_error/{}".format(error))
+            self.stats.inc_value("crawlera_fetch/response_error/{}".format(server_error))
             log_msg = (
-                "Error downloading <{} {}> (Original status: {}, Fetch API error message: {})"
+                "Error downloading <{} {}> (Original status: {}, Fetch API error message: {}). "
+                "Request ID: {}"
```
Comment (elacuesta) [inline crawlera_fetch/middleware.py:193]: Nitpick: what do you think about putting the request id inside the parenthesis? Or maybe removing them altoguether: `Error downloading <GET https://example.org> - Original status: 503, Fetch API error message: Server busy: too many outstanding requests, Request ID: None`

File: crawlera_fetch/middleware.py [hunk 3]
```diff
@@ -178,19 +181,23 @@ def process_response(self, request: Request, response: Response, spider: Spider)
...
             )
             log_msg = log_msg.format(
                 original_request.method,
                 original_request.url,
-                json_response["original_status"],
+                original_status,
                 message,
+                request_id
```
Comment (elacuesta) [inline crawlera_fetch/middleware.py]:
```suggestion
                request_id,
```

File: crawlera_fetch/middleware.py [hunk 4]
```diff
@@ -178,19 +181,23 @@ def process_response(self, request: Request, response: Response, spider: Spider)
                 logger.warning(log_msg)
                 return response
 
-        if json_response.get("crawlera_error"):
-            error = json_response["crawlera_error"]
-            message = json_response["body"]
+        server_error = json_response.get("crawlera_error") or json_response.get("error_code")
+        original_status = json_response.get("original_status", response.status)
```
[the remark this answers is not in the document]
Comment (edorofeev) [reply crawlera_fetch/middleware.py:185]: At this moment `original_status` can be missed for error responses (failed by timeout or something else)

File: crawlera_fetch/middleware.py [hunk 5]
```diff
@@ -199,26 +206,31 @@ def process_response(self, request: Request, response: Response, spider: Spider)
...
+            "crawlera_fetch/response_status_count/{}".format(original_status)
         )
 
         crawlera_meta["upstream_response"] = {
             "status": response.status,
             "headers": response.headers,
             "body": json_response,
         }
+        try:
+            resp_body = base64.b64decode(json_response["body"], validate=True)
```
[the remark this answers is not in the document]
Comment (edorofeev) [reply crawlera_fetch/middleware.py:216]: Tested:
- string compressed by zip
- bytes/unicode string
`binascii.Error` is catching all of them

File: crawlera_fetch/middleware.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (elacuesta) [reply crawlera_fetch/middleware.py:185]: I see. What I mean is, in that case, even if the upstream response fails, wouldn't the response from the Fetch API to the Scrapy process be a 200 one? I mean something like `Client <- (200) <- Fetch API <- (Timeout) <- Target site`, which would result in a "Error downloading, original status 200" message that I don't think would be reasonable to show.

[the remark this answers is not in the document]
Comment (edorofeev) [reply crawlera_fetch/middleware.py]: I've checked responses from latest version of API gateway and it's returning 520 if request failed for some reason. So, IMO it makes sense to use response.code from service.
But anyway I've updated the logic.
Could review it again please @elacuesta

**Pull request merged**